### PR TITLE
 Add Skyloong gk980 backlight keyboard.

### PR DESCRIPTION
**Pull request by NaturalZh**
# Skyloong GK980 backlight ansi keyboard

## Description
Skyloong GK980 backlight keyboard with 4 interchangeable encoders. It`s ansi layout.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/skyloong/gk980/bl/info.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        "debounce_type": "asym_eager_defer_pk"
+    },
+    "eeprom": {
+        "wear_leveling": {
+            "backing_size": 4096
+        }
+    },
+    "encoder": {
+        "rotary": [
+            {"pin_a": "A1", "pin_b": "A0", "resolution": 4},
```
Comment (waffle87) [inline keyboards/skyloong/gk980/bl/info.json:16]:
```suggestion
            {"pin_a": "A1", "pin_b": "A0"},
```

File: keyboards/skyloong/gk980/bl/ansi/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    case 0:
+        gpio_write_pin_low(MAC_PIN);
+        break;
+    case 1:
+        gpio_write_pin_high(MAC_PIN);
+        break;
+    default:
+        gpio_write_pin_low(MAC_PIN);
+
+    }
```
Comment (drashna) [inline keyboards/skyloong/gk980/bl/ansi/keymaps/via/keymap.c:71]: This entire block could be written as: 
```suggestion
    gpio_write_pin(MAC_PIN, layer_state_cmp(state, 1));
```

File: keyboards/skyloong/gk980/bl/bl.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+bool dis_breath = 0;
+uint8_t breath_level = 0;
+
+void suspend_power_down_kb() {
+    gpio_write_pin_low(MAC_PIN);
+    suspend_power_down_user();
+}
+
+void suspend_wakeup_init_kb(void) {
+    switch (get_highest_layer(layer_state)) {
```
Comment (drashna) [inline keyboards/skyloong/gk980/bl/bl.c:14]: I think you mean default_layer_state here. 

Also, the default_layer_state_set_user should be moved here, and to the kb function.

File: keyboards/skyloong/gk980/bl/bl.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+     case BL_BRTG:
+      if (record->event.pressed) {
+        if (dis_breath || !is_backlight_enabled()){
+            return false;
+        }
+            return true;
+        }
+
+     case TO(0):
```
Comment (drashna) [inline keyboards/skyloong/gk980/bl/bl.c:65]: These should be `DF`, not `TO`.  DF is meant for changing default layers.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/skyloong/gk980/bl/bl.c (hunk 3 above)
Comment (NaturalZh) [reply keyboards/skyloong/gk980/bl/bl.c]: OK, I do it.

File: keyboards/skyloong/gk980/bl/bl.c (hunk 4 above)
Comment (NaturalZh) [reply keyboards/skyloong/gk980/bl/bl.c:64]: It`s need to replace higher layers with a lower one. So TO is better than DF.

**Comment by NaturalZh** [comment]
What is it  means 'needs-wireless-code ' and how can I do? Would you please suggestion for it @drashna

**Comment by NaturalZh** [comment]
> This is the board, correct? https://skyloongtech.com/skyloong-gk980/
> 
> If so, then it needs to include the code for communicating with the wireless chip (but not the code for the firmware on the wireless chip) to be compliant with GPL licensing.
> 
> If that is not the case, could you confirm which board this is, specifically?

This is the keyboard of our alternative chipset solution featuring wireless communication capability, catering to users' diverse and personalized needs.   To meet the requirements of different customers, we typically develop a range of chipset solutions for keyboards.   The QMK keyboard is just one option;   currently, this chipset keyboard only supports USB communication.   If it's a QMK keyboard, relevant introductions and explanations are usually available on our website. @drashna

**Comment by tzarc** [comment]
As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.

As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

File: keyboards/skyloong/gk980/bl/bl.c (hunk 4 above)
Comment (NaturalZh) [reply keyboards/skyloong/gk980/bl/bl.c:64]: I modified it.

**Comment by NaturalZh** [comment]
> As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.
> 
> As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

OK, I modified it. Please check it again @tzarc

**Pull request closed without merging**